src/reliability.nim

```diff
@@ -0,0 +1,339 @@
+      if msgId in processed:
+        continue
+
+      # Process this message
+      for msg in rm.incomingBuffer:
+        if msg.messageId == msgId:
+          rm.addToHistory(msg.messageId)
+          if not rm.onMessageReady.isNil():
+            rm.onMessageReady(msg.messageId)
+          processed.incl(msgId)
+
+          # Add any dependent messages that might now be ready
+          if msgId in dependencies:
+            readyToProcess.add(dependencies[msgId])
+          break
+
+    rm.incomingBuffer = rm.incomingBuffer.filterIt(it.messageId notin processed)
+
+proc unwrapReceivedMessage*(
+    rm: ReliabilityManager, message: seq[byte]
+): Result[tuple[message: seq[byte], missingDeps: seq[SdsMessageID]], ReliabilityError] =
+  ## Unwraps a received message and processes its reliability metadata.
+  ##
+  ## Parameters:
+  ##   - message: The received message bytes
+  ##
+  ## Returns:
+  ##   A Result containing either tuple of (processed message, missing dependencies) or an error.
+  try:
+    let msg = deserializeMessage(message).valueOr:
+      return err(ReliabilityError.reDeserializationError)
+
+    if rm.bloomFilter.contains(msg.messageId):
+      return ok((msg.content, @[]))
+
+    rm.bloomFilter.add(msg.messageId)
+
+    # Update Lamport timestamp
+    rm.updateLamportTimestamp(msg.lamportTimestamp)
+
+    # Review ACK status for outgoing messages
+    rm.reviewAckStatus(msg)
+
+    var missingDeps: seq[SdsMessageID] = @[]
+    for depId in msg.causalHistory:
+      if not rm.bloomFilter.contains(depId):
+        missingDeps.add(depId)
+
+    if missingDeps.len == 0:
+      # Check if any dependencies are still in incoming buffer
+      var depsInBuffer = false
+      for bufferedMsg in rm.incomingBuffer:
+        if bufferedMsg.messageId in msg.causalHistory:
+          depsInBuffer = true
+          break
+
+      if depsInBuffer:
+        rm.incomingBuffer.add(msg)
+      else:
+        # All dependencies met, add to history
+        rm.addToHistory(msg.messageId)
+        rm.processIncomingBuffer()
+        if not rm.onMessageReady.isNil():
+          rm.onMessageReady(msg.messageId)
+    else:
+      rm.incomingBuffer.add(msg)
+      if not rm.onMissingDependencies.isNil():
+        rm.onMissingDependencies(msg.messageId, missingDeps)
+
+    return ok((msg.content, missingDeps))
+  except Exception:
+    error "Failed to unwrap message", msg = getCurrentExceptionMsg()
+    return err(ReliabilityError.reDeserializationError)
+
+proc markDependenciesMet*(
+    rm: ReliabilityManager, messageIds: seq[SdsMessageID]
+): Result[void, ReliabilityError] =
+  ## Marks the specified message dependencies as met.
+  ##
+  ## Parameters:
+  ##   - messageIds: A sequence of message IDs to mark as met.
+  ##
+  ## Returns:
+  ##   A Result indicating success or an error.
+  try:
+    # Add all messageIds to bloom filter
+    for msgId in messageIds:
+      if not rm.bloomFilter.contains(msgId):
+        rm.bloomFilter.add(msgId)
```

Comment:
Thinking that if we simplify the `incomingBuffer` to be a map of `messageId` to a `seq[SdsMessageID]` of missing dependencies (we can cache the full message contents elsewhere), we could directly mark these dependencies as met by removing them from the missingDependencies seq of each message in the incoming buffer. Processing the incoming buffer then becomes simply finding and processing the messages for which we now have an empty `missingDependencies` seq. WDYT?